components/authz-service/server/v2/rules_test.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/chef/automate/lib/grpc/grpctest"
+)
+
+func TestCreateRule(t *testing.T) {
+	ctx := context.Background()
+	cl, store, _, _ := setupRules(t)
+
+	// it's cumbersome to set this up, so we re-use it in a few of the following
+	// cases
+	apiConditions := []*api.Condition{
+		{
+			Type:   api.ProjectRuleConditionTypes_CHEF_ORGS,
+			Values: []string{"opscode"},
+		},
+	}
+	storageConditions := []storage.Condition{
+		{
+			Type:      storage.Node,
+			Attribute: storage.Organization,
+			Operator:  storage.MemberOf,
+			Value:     []string{"opscode"},
+		},
+	}
+
+	cases := []struct {
+		desc string
+		f    func(*testing.T)
+	}{
+		{"if the rule name is empty, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+				Id:         "empty-name",
+				Name:       "",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the rule id is empty, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+				Id:         "",
+				Name:       "empty id",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the rule id is invalid, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+				Id:         "no_underscores",
+				Name:       "any name",
+				ProjectId:  "foo",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the passed rule project id is invalid, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+				Id:         "any-name",
+				Name:       "any name",
+				ProjectId:  "foo bar",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if there are no conditions, returns 'invalid argument'", func(t *testing.T) {
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+				Id:        "foo",
+				Name:      "foo rule",
+				ProjectId: "bar",
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"if a rule with that id already exists, returns 'already exists'", func(t *testing.T) {
+			id := "foo-rule"
+			addRuleToStore(t, store, id, "my foo rule", storage.Node, "foo-project", storageConditions)
+
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+				Id:         id,
+				Name:       "my other foo",
+				ProjectId:  "bar",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.AlreadyExists, err)
+			assert.Nil(t, resp)
+		}},
+		// happy path
+		{"with valid rule data, returns no error and creates the rule in storage", func(t *testing.T) {
+			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
```

Comment:
Was curious whether this reduced form would work, but it fails. Any thoughts on why?
```
			ruleReq := &api.CreateRuleReq{
				Id:        "any-name",
				Name:      "any name",
				ProjectId: "foo",
				Type:      api.ProjectRuleTypes_NODE,
				Conditions: []*api.Condition{
					{
						Type:   api.ProjectRuleConditionTypes_CHEF_ORGS,
						Values: []string{"chef"},
					},
				},
			}
			resp, err := cl.CreateRule(ctx, ruleReq)
			assert.NoError(t, err)
			rule := api.ProjectRule(*ruleReq)
			assert.Equal(t, &api.CreateRuleResp{
				Rule: &rule,
			}, resp)
```

Reply:
Didn't dive in deeper, but with this:
```diff
diff --git a/components/authz-service/server/v2/rules_test.go b/components/authz-service/server/v2/rules_test.go
index 08a82b6c..c1be18cf 100644
--- a/components/authz-service/server/v2/rules_test.go
+++ b/components/authz-service/server/v2/rules_test.go
@@ -5,6 +5,7 @@ import (
 	"math/rand"
 	"testing"
 
+	"github.com/kylelemons/godebug/pretty"
 	cache "github.com/patrickmn/go-cache"
 	"github.com/stretchr/testify/assert"
 	"google.golang.org/grpc/codes"
@@ -103,7 +104,8 @@ func TestCreateRule(t *testing.T) {
 		}},
 		// happy path
 		{"with valid rule data, returns no error and creates the rule in storage", func(t *testing.T) {
-			resp, err := cl.CreateRule(ctx, &api.CreateRuleReq{
+
+			ruleReq := &api.CreateRuleReq{
 				Id:        "any-name",
 				Name:      "any name",
 				ProjectId: "foo",
@@ -114,22 +116,16 @@ func TestCreateRule(t *testing.T) {
 						Values: []string{"chef"},
 					},
 				},
-			})
+			}
+			actual, err := cl.CreateRule(ctx, ruleReq)
 			assert.NoError(t, err)
-			assert.Equal(t, &api.CreateRuleResp{
-				Rule: &api.ProjectRule{
-					Id:        "any-name",
-					Name:      "any name",
-					ProjectId: "foo",
-					Type:      api.ProjectRuleTypes_NODE,
-					Conditions: []*api.Condition{
-						{
-							Type:   api.ProjectRuleConditionTypes_CHEF_ORGS,
-							Values: []string{"chef"},
-						},
-					},
-				},
-			}, resp)
+			rule := api.ProjectRule(*ruleReq)
+			expected := &api.CreateRuleResp{
+				Rule: &rule,
+			}
+			if !assert.Equal(t, expected, actual) {
+				t.Log(pretty.Compare(expected, actual))
+			}
 		}},
 	}
```
...we get:
```
    --- FAIL: TestCreateRule/with_valid_rule_data,_returns_no_error_and_creates_the_rule_in_storage (0.00s)
        Error Trace:    rules_test.go:126
        Error:          Not equal:
                        expected: &v2.CreateRuleResp{Rule:(*v2.ProjectRule)(0xc000192690), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
                        actual  : &v2.CreateRuleResp{Rule:(*v2.ProjectRule)(0xc000192230), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}

                        Diff:
        Test:           TestCreateRule/with_valid_rule_data,_returns_no_error_and_creates_the_rule_in_storage
        rules_test.go:127:  {
              Rule: {
               Id: "any-name",
               ProjectId: "foo",
               Name: "any name",
               Type: 0,
               Conditions: [
                {
                 Type: 1,
                 Values: [
                  "chef",
                 ],
                 XXX_NoUnkeyedLiteral: {
                 },
                 XXX_unrecognized: [
                 ],
            -    XXX_sizecache: 8,
            +    XXX_sizecache: 0,
                },
               ],
               XXX_NoUnkeyedLiteral: {
               },
               XXX_unrecognized: [
               ],
            -  XXX_sizecache: 35,
            +  XXX_sizecache: 0,
              },
              XXX_NoUnkeyedLiteral: {
              },
              XXX_unrecognized: [
              ],
              XXX_sizecache: 0,
             }
```
So, the trivial answer is: the `XXX_sizecache` fields don't match. 😉

Reply:
Hmm... I did not see the expanded diff when I ran it, just this. Any idea why my output is different?
```
expected: &v2.CreateRuleResp{Rule:(*v2.ProjectRule)(0xc0001930a0), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
            
actual  : &v2.CreateRuleResp{Rule:(*v2.ProjectRule)(0xc000192bd0), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
                        
Diff:
```

Reply:
For more information please reread -- my comment includes the diff of what I had to do to see that, too ☝️

Reply:
🤔 Except it does not include the actual command; "diff --git..." is output... Am I missing something...?

Reply:
Sorry that was the output of `git diff`.